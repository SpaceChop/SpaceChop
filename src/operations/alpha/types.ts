import * as t from 'runtypes';
import Opacity from '../../types/Opacity';
import Param from '../../types/Param';

const AlphaConfig = t.Record({
  opacity: Opacity.Or(Param),
});

export type AlphaConfig = t.Static<typeof AlphaConfig>;
export default AlphaConfig;
