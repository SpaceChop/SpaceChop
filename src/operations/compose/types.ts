import * as t from 'runtypes';
import Arrangement from '../../types/Arrangement';
import Blending from '../../types/Blending';
import Composition from '../../types/Composition';
import Gravity from '../../types/Gravity';
import Offset from '../../types/Offset';
import Param from '../../types/Param';

const ComposeConfig = t.Record({
  source: t.String,
}).And(t.Partial({
  params: t.Dictionary(t.String, 'string'),
  preset: t.Union(t.String, Param),
  compose: t.Union(Composition, Param),
  arrange: t.Union(Arrangement, Param),
  gravity: t.Union(Gravity, Param),
  offset: Offset,
  blend: t.Union(Blending, Param),
}));

export type ComposeConfig = t.Static<typeof ComposeConfig>;
export default ComposeConfig;
