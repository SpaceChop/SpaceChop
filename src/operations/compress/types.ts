import * as t from 'runtypes';
import Param from '../../types/Param';

const Quality = t.Number.withConstraint(
  (n) => {
    if (n < 0 || n > 100) {
      return 'quality must be between 0 and 100';
    }
    if (n % 1 !== 0) {
      return 'quality must be non decimal number';
    }
    return true;
  },
);

const CompressConfig = t.Partial({
  quality: Quality.Or(Param),
  lossy: t.Boolean,
});

export type CompressConfig = t.Static<typeof CompressConfig>;
export default CompressConfig;
