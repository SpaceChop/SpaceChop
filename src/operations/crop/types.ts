import * as t from 'runtypes';
import Gravity from '../../types/Gravity';
import Offset from '../../types/Offset';
import Param from '../../types/Param';
import Percentage from '../../types/Percentage';
import PostiveNumber from '../../types/PositiveNumber';

const CropConfig = t.Partial({
  width: t.Union(PostiveNumber, Param, Percentage),
  height: t.Union(PostiveNumber, Param, Percentage),
  gravity: t.Union(Gravity, Param),
  offset: Offset,
}).withConstraint((n) => {
  if (!(n.width || n.height)) {
    return 'missing width or height';
  }
  return true;
});

export type CropConfig = t.Static<typeof CropConfig>;
export default CropConfig;
