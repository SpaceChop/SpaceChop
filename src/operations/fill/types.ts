import * as t from 'runtypes';
import Gravity from '../../types/Gravity';
import Param from '../../types/Param';
import PostiveNumber from '../../types/PositiveNumber';

const FillConfig = t.Record({
  width: t.Union(PostiveNumber, Param),
  height: t.Union(PostiveNumber, Param),
}).And(t.Partial({
  gravity: t.Union(Gravity, Param),
}));

export type FillConfig = t.Static<typeof FillConfig>;
export default FillConfig;
