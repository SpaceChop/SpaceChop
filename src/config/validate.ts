import console from '../lib/console';
import validate from '../types/validate';

let valid = true;

export default (config, throwError = false) => {
  try {
    validate(config);
    if (!valid) {
      valid = true;
      console.info('Config is valid again');
    }
    return true;
  } catch (err) {
    valid = false;
    console.error('\nThere is an error in /config.yml:');
    console.error('--------------------------------');
    console.error(err.message);
    console.error('--------------------------------');
    if (throwError) {
      throw new Error(err);
    }
  }
  return false;
};
