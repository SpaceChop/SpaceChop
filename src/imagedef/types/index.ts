import { Stream } from 'stream';
import ImageDefinition from '../../imagedef';
import AsyncBuffer from '../../lib/asyncBuffer';
import isGIF from './gif';
import isJPEG from './jpeg';
import isPNG from './png';
import isWEBP from './webp';

const tests = [
  isGIF,
  isJPEG,
  isPNG,
  isWEBP,
];

export default (stream: Stream): Promise<ImageDefinition> => {
  const asyncBuffer = new AsyncBuffer(stream);
  return new Promise((resolve, reject) => {
    tests.map((test) => {
      test(asyncBuffer)
        .then((definition) => {
          if (definition) {
            resolve(definition);
            asyncBuffer.destroy();
          }
        })
        .catch(reject);
    });
  });
};
