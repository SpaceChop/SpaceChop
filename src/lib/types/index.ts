import { Stream } from 'stream';
import ImageDefinition from '../../imagedef';
import isGIF from './gif';
import isJPEG from './jpeg';
import isPNG from './png';
import isWEBP from './webp';

const tests = [
  isGIF,
  isJPEG,
  isPNG,
  isWEBP,
];

export default (stream: Stream): Promise<any> => new Promise((resolve) => {
  const chunks = [];
  stream.on('data', (chunk) => {
    chunks.push(chunk);
  });
  stream.on('end', async () => {
    const buffer = Buffer.concat(chunks);
    for (const test of tests) {
      const result: ImageDefinition = test(buffer);
      if (result) {
        resolve(result);
        break;
      }
    }
  });
});
