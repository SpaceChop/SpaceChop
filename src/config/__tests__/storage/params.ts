import { Config } from '../../../types/Config';
import { validate } from '../../validate';

describe('validate storage', () => {
  it('should throw error on empty param in storage', () => {
    const config: Config = {
      paths: ['/:preset/:image'],
      sources: [{
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: ':image',
        },
      }],
      storage: {
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: '',
        },
      },
      presets: {},
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should throw error on empty path in storage', () => {
    const config: Config = {
      paths: ['/:preset/:uuid/:image'],
      sources: [{
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: ':uuid',
        },
      }],
      storage: {
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: '',
        },
      },
      presets: {},
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should throw error on missing param in storage', () => {
    const config: Config = {
      paths: ['/:preset/:uuid/:image'],
      sources: [{
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: ':uuid',
        },
      }],
      storage: {
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: '/:preset',
        },
      },
      presets: {},
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should throw error on missing param in path from storage', () => {
    const config: Config = {
      paths: ['/:preset/:uuid'],
      sources: [{
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: ':uuid',
        },
      }],
      storage: {
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: '/:preset/:image',
        },
      },
      presets: {},
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should not throw error on valid storage', () => {
    const config: Config = {
      paths: ['/:preset/:image'],
      sources: [{
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: ':image',
        },
      }],
      storage: {
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: ':image',
        },
      },
      presets: {},
    };
    expect(() => validate(config)).not.toThrow();
  });
});
