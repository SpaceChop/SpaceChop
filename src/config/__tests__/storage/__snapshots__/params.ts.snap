// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate storage should throw error on empty param in storage 1`] = `"storage.s3.path: Path cannot be empty"`;

exports[`validate storage should throw error on empty path in storage 1`] = `
"storage.s3.path: Path cannot be empty
params.image: Could not find any usage in sources, storage or paths"
`;

exports[`validate storage should throw error on missing param in path from storage 1`] = `"storage.s3.path: Missing param \\"image\\" in paths"`;

exports[`validate storage should throw error on missing param in storage 1`] = `
"storage.s3.path: Missing path parameters
params.image: Could not find any usage in sources, storage or paths"
`;
